pkgs/development/python-modules/pypass/default.nix

```diff
@@ -0,0 +1,53 @@
+{ buildPythonPackage
+, click
+, colorama
+, enum34
+, fetchFromGitHub
+, fetchgit
+, fetchPypi
+, pbr
+, pexpect
+, pythonAtLeast
+, pythonOlder
+, stdenv
+}:
+
+with stdenv.lib;
+
+# NOTE: pypass can also be used as an application, but probably the most
+# important usecase is as a library. So, let's use buildPythonPackage and
+# support any Python version instead of defining it as an application with
+# buildPythonApplication.
+buildPythonPackage rec {
+  pname = "pypass";
+  version = "0.2.1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1nm4mj7pd7gz4ghic6b3wrnd1b59hd1f0axavdabfl79wy511l7r";
+  };
+
+  # Remove enum34 requirement if Python >= 3.4
+  postPatch = optionalString (pythonAtLeast "3.4") ''
+    substituteInPlace requirements.txt --replace "enum34" ""
+  '';
+
+  propagatedBuildInputs = [
+    click
+    colorama
+    pexpect
+  ] ++ optionals (pythonOlder "3.4") enum34;
+
+  # It seemed difficult to get the tests working..
```

Comment:
Can you add more details what is difficult?

Reply:
The tests require, for instance, gpg and then I get errors like:

```
gpg: Fatal: can't create directory '/homeless-shelter/.gnupg': No such file or directory
```

Do you have ideas how to solve this? Should I create some temporary directory and set it as the home directory in the check phase?

Similary, the tests seem to require git for some reason and then i get errors that some basic information needs to be set. I set them in `preCheck` as follows:

```
    git config --global user.email "you@example.com"
    git config --global user.name "Your Name"
```

But then I get an error:

```
error: could not lock config file /homeless-shelter/.gitconfig: No such file or directory
```

So, a similar home directory issue.

Reply:
I got a bit forward by setting `HOME=$TEMP`. Still some failures, but I'll look into those.

For completeness/reference, here are the current failures I'm having:

```
....FInitialized empty Git repository in /build/tmp8guynk9_/.git/
[master (root-commit) bd396c5] Add current contents of password store.
 1 file changed, 1 insertion(+)
 create mode 100644 .gpg-id
[master 838a54c] Configure git repository for gpg file diff.
 1 file changed, 1 insertion(+)
 create mode 100644 .gitattributes
.gpg: directory '/build/.gnupg' created
gpg: keybox '/build/.gnupg/pubring.kbx' created
gpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
F.Initialized empty Git repository in /build/tmp8uv5098j/.git/
[master (root-commit) 5e11da8] "testcommit"
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 test_git_init_clone.gpg
Initialized empty Git repository in /build/tmpt7p_6oza/.git/
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (3/3), 204 bytes | 204.00 KiB/s, done.
From /build/tmp8uv5098j
 * branch            master     -> FETCH_HEAD
 * [new branch]      master     -> origin/master
.gpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
FF......gpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
F..gpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
F.gpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
gpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
gpg: can't open '/build/tmp_yfa4k51/A/B/C/D/hello.com.gpg': No such file or directory
gpg: decrypt_message failed: No such file or directory
Fgpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
gpg: can't open '/build/tmp18k0rx7k/hello.com.gpg': No such file or directory
gpg: decrypt_message failed: No such file or directory
Fgpg: 5C5833E3: skipped: No public key
gpg: [stdin]: encryption failed: No public key
gpg: can't open '/build/tmpq1yf8bt4/hello.com.gpg': No such file or directory
gpg: decrypt_message failed: No such file or directory
F..Initialized empty Git repository in /build/tmprfzffkiq/.git/
[master (root-commit) 5e11da8] "testcommit"
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 test_git_init_clone.gpg
Initialized empty Git repository in /build/tmpn8_c8yqm/.git/
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (3/3), 204 bytes | 204.00 KiB/s, done.
From /build/tmprfzffkiq
 * branch            master     -> FETCH_HEAD
 * [new branch]      master     -> origin/master
.
======================================================================
FAIL: test_find (pypass.tests.test_command.TestCommand)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_command.py", line 280, in test_find
    self.assertIsNotNone(re.search(expected_regex, find_result.output))
AssertionError: unexpectedly None

======================================================================
FAIL: test_grep (pypass.tests.test_command.TestCommand)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_command.py", line 287, in test_grep
    self.assertEqual(
AssertionError: '' != 'grep_test.com:\nGREPME\n'
+ grep_test.com:
+ GREPME


======================================================================
FAIL: test_insert_and_show (pypass.tests.test_command.TestCommand)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_command.py", line 94, in test_insert_and_show
    self.assertTrue(
AssertionError: False is not true

======================================================================
FAIL: test_ls (pypass.tests.test_command.TestCommand)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_command.py", line 155, in test_ls
    self.assertIsNotNone(re.search(expected_regex, ls_result.output))
AssertionError: unexpectedly None

======================================================================
FAIL: test_show_clip (pypass.tests.test_command.TestCommand)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_command.py", line 123, in test_show_clip
    self.assertEqual(
AssertionError: 'Error: clip_test is not in the password store.\n' != 'Copied clip_test to clipboard.\n'
- Error: clip_test is not in the password store.
+ Copied clip_test to clipboard.


======================================================================
FAIL: test_encrypt_decrypt (pypass.tests.test_passwordstore.TestPasswordStore)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_passwordstore.py", line 80, in test_encrypt_decrypt
    self.assertTrue(
AssertionError: False is not true

======================================================================
FAIL: test_get_decrypted_password_deeply_nested (pypass.tests.test_passwordstore.TestPasswordStore)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_passwordstore.py", line 129, in test_get_decrypted_password_deeply_nested
    self.assertEqual(
AssertionError: 'Alice' != None

======================================================================
FAIL: test_get_decrypted_password_only_password (pypass.tests.test_passwordstore.TestPasswordStore)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_passwordstore.py", line 117, in test_get_decrypted_password_only_password
    self.assertEqual(
AssertionError: 'ELLO' != None

======================================================================
FAIL: test_get_decrypted_password_specific_entry (pypass.tests.test_passwordstore.TestPasswordStore)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/pypass-0.2.1/pypass/tests/test_passwordstore.py", line 96, in test_get_decrypted_password_specific_entry
    self.assertEqual(
AssertionError: 'ELLO' != None

----------------------------------------------------------------------
Ran 28 tests in 0.254s

FAILED (failures=9)
```

Reply:
The tests failing shouldn't be a blocker. You should test some functionality though, and as Python maintainers would have recommended `pythonImportsCheck`.

Reply:
I got the tests working now.